rai/node/node.cpp

```diff
@@ -335,154 +474,1199 @@ void rep_query (rai::node & node_a, T const & peers_a)
+{
+	if (options.size () == 0)
+	{
+		return std::vector<std::vector<uint64_t *>> (1, options); // ???
+	}
+	// round up
+	size_t block_chunks = ((options.size () - 1) / 64) + 1;
+	// leave space for b
+	size_t chunks = (options.size () / 64) + 1;
+	size_t check_bits = check_size * 64;
+	// Solve Ax=b for x
+	// Elements are 1 bit and overflow is expected
+	std::vector<std::vector<uint64_t>> matrix (check_bits, std::vector<uint64_t> (chunks, 0));
+	// Fill the matrix
+	for (size_t bit = 0; bit < check_bits; bit++)
+	{
+		for (size_t i = 0; i < options.size (); i++)
+		{
+			matrix[bit][i / 64] |= ((options[i][bit / 64] >> (bit % 64)) & 1) << (i % 64);
+		}
+		matrix[bit][chunks - 1] |= ((xor_check[bit / 64] >> (bit % 64)) & 1) << (options.size () % 64);
+	}
+	// Put matrix in upper triangular form
+	std::vector<size_t> row_to_col;
+	row_to_col.reserve (check_bits);
+	std::vector<size_t> free_cols;
+	size_t pivot_row = 0;
+	for (size_t col = 0; col < options.size (); col++)
+	{
+		size_t chunk = col / 64;
+		uint64_t chunk_mask = 1UL << (col % 64);
+		bool found_pivot = false;
+		for (size_t row = pivot_row; row < matrix.size (); row++)
+		{
+			if ((matrix[row][chunk] & chunk_mask) != 0)
+			{
+				swap (matrix[row], matrix[pivot_row]);
+				found_pivot = true;
+				break;
+			}
+		}
+		if (found_pivot)
+		{
+			assert (row_to_col.size () == pivot_row);
+			row_to_col.push_back (col);
+			for (size_t row = pivot_row + 1; row < matrix.size (); row++)
+			{
+				if ((matrix[row][chunk] & chunk_mask) != 0)
+				{
+					for (size_t chunk = 0; chunk < chunks; chunk++)
+					{
+						matrix[row][chunk] ^= matrix[pivot_row][chunk];
+					}
+				}
+			}
+			pivot_row += 1;
+		}
+		else
+		{
+			// We still continue because this prioritizes high-stake conflicts
+			if (free_cols.size () < possible_cap_log2)
+			{
+				free_cols.push_back (col);
+			}
+		}
+	}
+	size_t b_chunk = chunks - 1;
+	uint64_t b_chunk_mask = 1UL << (options.size () % 64);
+	for (uint64_t row = pivot_row; row < matrix.size (); row++)
+	{
+		if ((matrix[row][b_chunk] & b_chunk_mask) != 0)
+		{
+			return std::vector<std::vector<uint64_t *>> (); // no solution
+		}
+	}
+	std::vector<std::vector<uint64_t>> free_cols_flip;
+	for (size_t i = 0; i < free_cols.size (); i++)
+	{
+		size_t free_col = free_cols[i];
+		std::vector<uint64_t> mask (block_chunks, 0);
+		mask[free_col / 64] |= 1UL << (free_col % 64);
+		free_cols_flip.push_back (mask);
+	}
+	std::vector<uint64_t> base_cols (chunks, 0);
+	base_cols[b_chunk] = b_chunk_mask;
+	size_t i = pivot_row;
+	while (i != 0)
+	{
+		i--;
+		size_t col = row_to_col[i];
+		size_t col_chunk = col / 64;
+		uint64_t col_chunk_mask = 1UL << (col % 64);
+		for (size_t f = 0; f < free_cols_flip.size (); f++)
+		{
+			uint64_t chunks_xor = 0;
+			for (size_t chunk = 0; chunk < matrix[i].size (); chunk++)
+			{
+				chunks_xor ^= matrix[i][chunk] & free_cols_flip[f][chunk];
+			}
+			if (__builtin_parityl (chunks_xor))
```

Comment:
Maybe add msvc compatibility to the PR's TODO list ? Possibly a compiler specific macro that does __builtin_popcount (gcc) or __popcnt (msvc) with an LSB mask.